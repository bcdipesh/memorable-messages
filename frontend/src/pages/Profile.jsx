import { useContext, useEffect, useState } from "react";
import { useForm } from "react-hook-form";

import AuthContext from "@/contexts/authContext/AuthContext";

import { yupResolver } from "@hookform/resolvers/yup";
import * as yup from "yup";

import MemorableMessagesApi from "@/apis/memorableMessages/memorableMessagesApi";
import { Button } from "@/components/ui/button";
import {
  Card,
  CardContent,
  CardDescription,
  CardFooter,
  CardHeader,
  CardTitle,
} from "@/components/ui/card";
import {
  Form,
  FormControl,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from "@/components/ui/form";
import { Input } from "@/components/ui/input";
import { Skeleton } from "@/components/ui/skeleton";
import { Tabs, TabsContent, TabsList, TabsTrigger } from "@/components/ui/tabs";
import { useToast } from "@/components/ui/use-toast";

/**
 * Profile component renders form managing user account details and password.
 *
 * @returns {React.JSX.Element} Profile component UI.
 */
const Profile = () => {
  const [isLoading, setIsLoading] = useState(false);
  const [username, setUsername] = useState("");
  const [email, setEmail] = useState("");
  const { toast } = useToast();
  const { handleLogout } = useContext(AuthContext);

  // Form setup for user details
  const userDetailsForm = useForm({
    resolver: yupResolver(
      yup.object({
        username: yup.string().required(),
        email: yup.string().required(),
      }),
    ),
    defaultValues: {
      username: "",
      email: "",
    },
    values: {
      username,
      email,
    },
  });

  // Form setup for user password
  const userPasswordForm = useForm({
    resolver: yupResolver(
      yup.object({
        currentPassword: yup.string().required(),
        newPassword: yup.string().required(),
      }),
    ),
    defaultValues: {
      currentPassword: "",
      newPassword: "",
    },
  });

  /**
   * Display a toast notification with optional variant
   * @param title - Title of the notification
   * @param description - Description of the notification
   * @param variant - Optional variant (default, success, destructive)
   */
  const showToast = (title, description, variant = "default") =>
    toast({
      variant,
      title,
      description,
    });

  useEffect(() => {
    // Set document title
    document.title = "Memorable Messages | Profile";

    /**
     * Fetch user details and update state variables.
     */
    const getUserDetails = async () => {
      setIsLoading(true);
      try {
        const resp = await MemorableMessagesApi.getCurrentUserDetails();
        setIsLoading(false);
        setUsername(resp.username);
        setEmail(resp.email);
      } catch (err) {
        setIsLoading(false);
        if (err.statusCode === 401) {
          handleLogout();
          showToast(
            "Oops! It seems your session has expired.",
            "Please log back in using your username and password and try again.",
            "destructive",
          );
        }
      }
    };

    getUserDetails();
  }, []);

  /**
   * Handle form submission for updating user details.
   * @param {object} data - Form data.
   * @param {Event} e - Form submission event.
   */
  const onSubmit = async (data, e) => {
    e.preventDefault();

    try {
      const resp = await MemorableMessagesApi.updateUserDetails(data);
      setUsername(resp.username);
      setEmail(resp.email);
    } catch (err) {
      if (err.statusCode === 400) {
        showToast(
          "Uh oh! Something went wrong.",
          "We couldn't process your request. Please double-check your information and try again.",
          "destructive",
        );
      } else if (err.statusCode === 401) {
        handleLogout();
        showToast(
          "Oops! It seems your session has expired.",
          "Please log back in using your username and password and try again.",
          "destructive",
        );
      }
    }
  };

  /**
   * Handle form submission for updating user password.
   * @param {object} data - Form data.
   * @param {Event} e - Form submission event.
   */
  const updatePassowrd = async (data, e) => {
    e.preventDefault();
    try {
      // Log in with the current password
      await MemorableMessagesApi.login(username, data.currentPassword);

      // Update user details with the new password
      await MemorableMessagesApi.updateUserDetails({
        username,
        email,
        password: data.newPassword,
      });
      showToast(
        "Password updated!",
        "Your password has been successfully changed. Remember to keep it safe!",
      );

      // Log out after updating the password
      handleLogout();
    } catch (err) {
      if (err.statusCode === 400) {
        showToast(
          "Uh oh! Something went wrong.",
          "We couldn't process your request. Please double-check your information and try again.",
          "destructive",
        );
      } else if (err.statusCode === 401) {
        showToast(
          "Uh oh! Something went wrong.",
          "The password you entered doesn't match your current password. Please try again.",
          "destructive",
        );
      }
    }
  };

  return (
    <Tabs defaultValue="account" className="h-1/2">
      <TabsList className="grid w-full grid-cols-2">
        <TabsTrigger value="account">Account</TabsTrigger>
        <TabsTrigger value="password">Password</TabsTrigger>
      </TabsList>
      {/* User details */}
      <TabsContent value="account">
        <Card>
          <CardHeader>
            <CardTitle>Account</CardTitle>
            <CardDescription>
              Make changes to your account here. Click save when you're done.
            </CardDescription>
          </CardHeader>
          <CardContent>
            {isLoading && (
              <div className="user-details-loading-skeleton flex flex-col space-y-4">
                <Skeleton className="h-5 w-16" />
                <Skeleton className="h-10 w-full" />

                <Skeleton className="h-5 w-16" />
                <Skeleton className="h-10 w-full" />
              </div>
            )}{" "}
            {!isLoading && (
              <Form {...userDetailsForm}>
                <form
                  id="userDetailsForm"
                  onSubmit={userDetailsForm.handleSubmit(onSubmit)}
                  className="space-y-4"
                >
                  <FormField
                    control={userDetailsForm.control}
                    name="username"
                    render={({ field }) => (
                      <FormItem>
                        <FormLabel>Username</FormLabel>
                        <FormControl>
                          <Input placeholder="username" {...field} />
                        </FormControl>
                        <FormMessage />
                      </FormItem>
                    )}
                  />

                  <FormField
                    control={userDetailsForm.control}
                    name="email"
                    render={({ field }) => (
                      <FormItem>
                        <FormLabel>Email</FormLabel>
                        <FormControl>
                          <Input
                            type="email"
                            placeholder="name@example.com"
                            {...field}
                          />
                        </FormControl>
                        <FormMessage />
                      </FormItem>
                    )}
                  />
                </form>
              </Form>
            )}
          </CardContent>
          <CardFooter>
            <Button form="userDetailsForm" type="submit">
              Save changes
            </Button>
          </CardFooter>
        </Card>
      </TabsContent>

      {/* Update password */}
      <TabsContent value="password">
        <Card>
          <CardHeader>
            <CardTitle>Password</CardTitle>
            <CardDescription>
              Change your password here. After saving, you'll be logged out.
            </CardDescription>
          </CardHeader>
          <CardContent className="space-y-2">
            <Form {...userPasswordForm}>
              <form
                id="userPasswordForm"
                onSubmit={userPasswordForm.handleSubmit(updatePassowrd)}
                className="space-y-4"
              >
                <FormField
                  control={userPasswordForm.control}
                  name="currentPassword"
                  render={({ field }) => (
                    <FormItem>
                      <FormLabel>Current password</FormLabel>
                      <FormControl>
                        <Input type="password" {...field} />
                      </FormControl>
                      <FormMessage />
                    </FormItem>
                  )}
                />

                <FormField
                  control={userPasswordForm.control}
                  name="newPassword"
                  render={({ field }) => (
                    <FormItem>
                      <FormLabel>New password</FormLabel>
                      <FormControl>
                        <Input type="password" {...field} />
                      </FormControl>
                      <FormMessage />
                    </FormItem>
                  )}
                />
              </form>
            </Form>
          </CardContent>
          <CardFooter>
            <Button type="submit" form="userPasswordForm">
              Update password
            </Button>
          </CardFooter>
        </Card>
      </TabsContent>
    </Tabs>
  );
};

export default Profile;
