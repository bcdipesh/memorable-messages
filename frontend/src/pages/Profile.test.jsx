import React from "react";
import { BrowserRouter } from "react-router-dom";

import { render } from "@testing-library/react";

import Profile from "@/pages/Profile";

test("Renders without crashing", function () {
  render(
    <BrowserRouter>
      <Profile />
    </BrowserRouter>,
  );
});

test("Matches snapshot", function () {
  const { asFragment } = render(
    <BrowserRouter>
      <Profile />
    </BrowserRouter>,
  );
  expect(asFragment()).toMatchSnapshot();
});
