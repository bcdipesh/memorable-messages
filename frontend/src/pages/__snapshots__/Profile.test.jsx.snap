// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="h-1/2"
    data-orientation="horizontal"
    dir="ltr"
  >
    <div
      aria-orientation="horizontal"
      class="h-10 items-center justify-center rounded-md bg-muted p-1 text-muted-foreground grid w-full grid-cols-2"
      data-orientation="horizontal"
      role="tablist"
      style="outline: none;"
      tabindex="0"
    >
      <button
        aria-controls="radix-:r5:-content-account"
        aria-selected="true"
        class="inline-flex items-center justify-center whitespace-nowrap rounded-sm px-3 py-1.5 text-sm font-medium ring-offset-background transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=active]:bg-background data-[state=active]:text-foreground data-[state=active]:shadow-sm"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-state="active"
        id="radix-:r5:-trigger-account"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Account
      </button>
      <button
        aria-controls="radix-:r5:-content-password"
        aria-selected="false"
        class="inline-flex items-center justify-center whitespace-nowrap rounded-sm px-3 py-1.5 text-sm font-medium ring-offset-background transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=active]:bg-background data-[state=active]:text-foreground data-[state=active]:shadow-sm"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-state="inactive"
        id="radix-:r5:-trigger-password"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Password
      </button>
    </div>
    <div
      aria-labelledby="radix-:r5:-trigger-account"
      class="mt-2 ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2"
      data-orientation="horizontal"
      data-state="active"
      id="radix-:r5:-content-account"
      role="tabpanel"
      style="animation-duration: 0s;"
      tabindex="0"
    >
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm"
      >
        <div
          class="flex flex-col space-y-1.5 p-6"
        >
          <h3
            class="text-2xl font-semibold leading-none tracking-tight"
          >
            Account
          </h3>
          <p
            class="text-sm text-muted-foreground"
          >
            Make changes to your account here. Click save when you're done.
          </p>
        </div>
        <div
          class="p-6 pt-0"
        >
          <div
            class="user-details-loading-skeleton flex flex-col space-y-4"
          >
            <div
              class="animate-pulse rounded-md bg-muted h-5 w-16"
            />
            <div
              class="animate-pulse rounded-md bg-muted h-10 w-full"
            />
            <div
              class="animate-pulse rounded-md bg-muted h-5 w-16"
            />
            <div
              class="animate-pulse rounded-md bg-muted h-10 w-full"
            />
          </div>
           
        </div>
        <div
          class="flex items-center p-6 pt-0"
        >
          <button
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
            form="userDetailsForm"
            type="submit"
          >
            Save changes
          </button>
        </div>
      </div>
    </div>
    <div
      aria-labelledby="radix-:r5:-trigger-password"
      class="mt-2 ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2"
      data-orientation="horizontal"
      data-state="inactive"
      hidden=""
      id="radix-:r5:-content-password"
      role="tabpanel"
      tabindex="0"
    />
  </div>
</DocumentFragment>
`;
